intersphinx xref

nstarman opened the pull request:
Signed-off-by: Nathaniel Starkman (@nstarman) <nstarkman@protonmail.com>

- Add section on numpydoc xref to v1.
- preload the xref ignore set. This can be modified or replaced in packages with standard set operations.
- Add ``numpydoc_xref_aliases_astropy_glossary``, an xref mapping that can be added to the ``numpydoc_xref_aliases`` in packages by dictionary updating. This is designed for easy inclusion of astropy's glossary.
- Add ``numpydoc_xref_aliases_astropy_physical_type``, an xref mapping that can be added to the ``numpydoc_xref_aliases`` in packages by dictionary updating. This is designed for easy inclusion of astropy's physical-type x-refs.
- Add ``numpydoc_xref_astropy_aliases``, an xref ChainMap of the above mappings that can be added to the ``numpydoc_xref_aliases`` in packages by dictionary updating. This is designed for easy inclusion of ALL of astropy's intersphinx offerings.

Fixes #39
Fixes #20
Request Review : @bsipocz @pllim @saimn @astrofrog

nstarman commented on sphinx_astropy/conf/v1.py:
```diff
@@ -88,6 +89,42 @@ def check_sphinx_version(expected_version):
+# aliases/shortcuts used when specifying the types of parameters.
+# Numpy provides some defaults
+# https://github.com/numpy/numpydoc/blob/b352cd7635f2ea7748722f410a31f937d92545cc/numpydoc/xref.py#L62-L94
+# numpydoc_xref_aliases = {}
```
maybe this should be uncommented?

nstarman replied:
yup.

nstarman commented on sphinx_astropy/conf/v1.py:
```diff
@@ -88,6 +89,42 @@ def check_sphinx_version(expected_version):
+            val = f":ref:`:ref: '{ptype}' <astropy:{ptype}>`"   # <= intersphinxed.
+            numpydoc_xref_aliases_astropy_physical_type[key] = val
+    
+    del ptypes, key, val
```
del import as well?

nstarman replied:
Done.

nstarman commented on sphinx_astropy/conf/v1.py:
```diff
@@ -88,6 +89,42 @@ def check_sphinx_version(expected_version):
+# convenience.
+numpydoc_xref_ignore = {
+    'type', 'optional', 'default', 'or', 'of', 'method', 'instance', "like",
+    "class", 'subclass', "keyword-only", "default", "thereof"
```
add more from Astropy's conf.py? mixin? Numpy stuff?

nstarman replied:
rm "like" ?

nstarman replied:
Done.

nstarman commented on setup.cfg:
```diff
@@ -28,6 +28,7 @@ install_requires =
 	sphinx-automodapi
 	sphinx-gallery
 	pillow
+    astropy
```
add spaces

nstarman commented on sphinx_astropy/conf/v1.py:
```diff
@@ -88,6 +89,42 @@ def check_sphinx_version(expected_version):
+# Numpy provides some defaults
+# https://github.com/numpy/numpydoc/blob/b352cd7635f2ea7748722f410a31f937d92545cc/numpydoc/xref.py#L62-L94
+# numpydoc_xref_aliases = {}
+
```
```suggestion

# aliases to Astropy's physical types. In packages these can be turned on with
# ``numpydoc_xref_aliases.update(numpydoc_xref_aliases_astropy_physical_type)``
```

nstarman commented:
https://github.com/astropy/astropy/pull/11684 shows that this will not break astropy due to circular imports.
I don't think this will be super testable until astropy 4.3 is the 'stable' version in the intersphinx connection. @adrn and I have a PR in `gala` to test this when that happens. https://github.com/astropy/astropy/pull/11684 is likewise set up for testing this PR in astropy itself (but that has added complications due to intersphinx).

nstarman commented on sphinx_astropy/conf/v1.py:
```diff
@@ -88,6 +89,50 @@ def check_sphinx_version(expected_version):
+# aliases/shortcuts used when specifying the types of parameters.
+# Numpy provides some defaults
+# https://github.com/numpy/numpydoc/blob/b352cd7635f2ea7748722f410a31f937d92545cc/numpydoc/xref.py#L62-L94
+numpydoc_xref_aliases = {
```
Add more stuff from astropy's config?

nstarman replied:
I'll hold off until there is a need for more.

nstarman commented on sphinx_astropy/conf/v1.py:
```diff
@@ -88,6 +90,84 @@ def check_sphinx_version(expected_version):
+    numpydoc_xref_aliases_astropy_glossary = {
+        # general
+        "-like": ":term:`astropy:-like`",
+        # "number": ":term:`number`",
+        # "writable": ":term:`writable file-like object`",
+        # "readable": ":term:`readable file-like object`",
```
how many of these should be included?

nstarman replied:
I'm not including any until there is need / I see this pop up in an affiliate package's conf.py.

saimn commented:
I'm not convinced about the cross-reference/dependency with astropy, this may cause problems in the future. Could we put a static list instead ? I guess the list doesn't change often so we could avoid the dynamic generation.
Also numpydoc xref should be disabled by default (which is the case here I think), and we could just provide `numpydoc_xref_aliases_astropy_glossary` and `numpydoc_xref_aliases_astropy_physical_type` (no matter the astropy version) and tell people to include what they need in `numpydoc_xref_aliases`.

nstarman commented:
> I'm not convinced about the cross-reference/dependency with astropy, this may cause problems in the future.

Do you mean having astropy be a run-time requirement of this package? I was initially concerned (hence astropy/astropy#11684) as I wasn't sure that the local package (astropy) got installed before the documentation requirements, but since it does, there shouldn't be any issues unless the installation order changes.

> Could we put a static list instead ? I guess the list doesn't change often so we could avoid the dynamic generation.

The issue then is that any time astropy wants to update the list, sphinx-astropy needs a new release. Dynamic generation obviates the maintainer burden.

> Also numpydoc xref should be disabled by default (which is the case here I think)

Agreed. It is here. I have a comment explaining how to turn it on.

> and we could just provide `numpydoc_xref_aliases_astropy_glossary` and `numpydoc_xref_aliases_astropy_physical_type` (no matter the astropy version) and tell people to include what they need in `numpydoc_xref_aliases`.

I don't think I understand what you're suggesting here.

One large danger with adding sphinx links to the numpydoc xref regardless of astropy version is that they can cause the docs to fail, depending on the astropy version. I don't think NOT including them will fail a numpydoc xref unless the the docs are built with an earlier version of astropy than is supported in the code.

nstarman commented:
ping @astrofrog @pllim @saimn @bsipocz.
I've cleaned up the dynamic generation and now that https://github.com/astropy/astropy/pull/11690 is in, the intersphinx links in ``numpydoc_xref_aliases_astropy_glossary``, and ``numpydoc_xref_aliases_astropy_physical_type`` all programagically work in Astropy without any modifications! It should be convenient for both Astropy and affiliate packages. The Astropy side is ready with https://github.com/astropy/astropy/pull/11684.

astrofrog commented on sphinx_astropy/conf/v1.py:
```diff
@@ -88,6 +90,77 @@ def check_sphinx_version(expected_version):
+# ``numpydoc_xref_aliases.update(numpydoc_xref_aliases_astropy_physical_type)``
+# (if astropy is in the intersphinx mapping).
+numpydoc_xref_aliases_astropy_physical_type = {}  # works even if older astropy
+if float(astropy.__version__[:3]) >= 4.3:
```
We can easily keep Astropy as an optional dependency for this package by defining ASTROPY_INSTALLED further up and only do the following check if it is?

nstarman replied:
Thanks @astrofrog. That seems an excellent solution!
I'll make the necessary changes.

astrofrog reviewed (changes requested):
I am in agreement with @saimn that we shouldn't have astropy as a dependency here - instead we should just make use of it if it is installed but not otherwise. For most packages that use this astropy will be a required dependency anyway, and for astropy core we will avoid installing a stable astropy into a development build of the astropy docs.

To be clear, is there any risk of this breaking anything which would warrant a v2 config? I suspect not since this is opt-in, but just wanted to check.

nstarman commented:
@saimn, I've incorporated the @astrofrog's suggestion to make Astropy an optional dependency. The xref dictionaries are empty if Astropy is not installed or <v4.3 and full otherwise. Does this address the points you raised about cross-dependencies and working around that by static lists?

saimn reviewed (commented):
Thanks @nstarman, yes I agree the approach seems reasonable now. Just needs a small change.

nstarman commented on sphinx_astropy/conf/v1.py:
```diff
@@ -88,6 +96,77 @@ def check_sphinx_version(expected_version):
+# convenience.
+numpydoc_xref_ignore = {
+    "type", "optional", "default", "or", "of", "method", "instance",
+    "class", "subclass", "keyword-only", "default", "thereof",
```
Great. I'll push the change.

nstarman replied:
Done!

saimn reviewed (approved):
Thanks @nstarman, sounds good now.

@astrofrog - Do you want to have another look ?

nstarman commented:
I've squashed some unnecessary commits.
Also, I changed the version check from ``float(astropy.__version__[:3]) >= 4.3`` to ``minversion(astropy, "4.3")``. Much cleaner.

saimn commented:
Thanks @nstarman !

nstarman commented:
Thanks for the reviews!
Now to get this into Astropy. But I guess that will require a new release of sphinx-astropy? Or is there a nightly release or something?

pllim commented:
Re: release -- Probably needs a new release but... @saimn , looks like we never set up a auto publish workflow here. Hmm!

saimn commented:
Yep, will try to set it up in the coming days.

pllim commented:
@nstarman , this is now in https://pypi.org/project/sphinx-astropy/1.4/

nstarman commented:
Awesome! Thanks.
Now I can get this in for Astropy main.

pllim commented:
Many thanks to @saimn ! 😸
